### ISSUE TITLE
0002930:
Dovecot MySQL IMAP & Postfix MySQL support

### ISSUE THREAD

**Issue by Gloirin** [tine20/Tine-2.0-Open-Source-Groupware-and-CRM#1543]
**Reported by raghnall on 17 Aug 2010 06:19**

**Version:** git master

These files contain support for Dovecot IMAP server with Postfix set to use the dovecot database. 
Also contains a Dummy Driver for SMTP as the Dovecot IMAP driver handles both IMAP and SMTP.
Dummy Driver works for IMAP or SMTP. 
This allows an SMTP driver to function as both IMAP and SMTP like the Dovecot driver does.

**Additional information:** Files:
Setup/js/EmailPanel.js
Tinebase/EmailUser.php
Tinebase/EmailUser/Dummy.php
Tinebase/EmailUser/Imap/Dovecot.php

**Comment by Gloirin** [comment]
Depends on #1729

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 17 Aug 2010 09:58**

@raghnall: thx! could you please send us a signed CLA? we need it to include your code:

http://www.tine20.org/wiki/index.php/Become_a_developer_/_CLA

and send it to lars or me (email addresses can be found in the forum).

@lars: could you please review the code?

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 17 Aug 2010 18:57**

Looks very good so far.

But it would be better to keep SMTP and IMAP settings in different classes, even they can&#039;t work without each other. We also like to restructure the internal logic of the email setup stuff, which will require the classes to be separated.

Can you rework the classes this way? Then I will integrate them into Tine 2.0.

Lars

**Comment by Gloirin** [comment]
**Comment posted by raghnall on 18 Aug 2010 00:51**

OK, Updated to work separately.

File List:
Setup/js/EmailPanel.js
Tinebase/EmailUser.php
Tinebase/EmailUser/Dummy.php
Tinebase/EmailUser/Dovecot.php
Tinebase/EmailUser/Imap/Dovecot.php
Tinebase/EmailUser/Smtp/Dovecot.php

Dummy is no longer needed, but I still use it when figuring out what parts need to go where in the respective drivers.

Tinebase/EmailUser/Dovecot.php is an abstract class that contains a few functions that are needed by both the IMAP and SMTP drivers. Otherwise they are separate.

The Setup EmailPanel now contains identical settings for IMAP and SMTP. If both are configured IMAP wins. (or at least it should)

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 20 Aug 2010 08:07**

I spotted another problem. But maybe we must live with it.

The sql table dovecot_users contains the field email which must be unique and not null.

This means that we can&#039;t remove the email address of an imap account without removing the whole imap account. If we could move the email address to a separate table, then we could have the imap account informations separated from the smtp settings.

We have this situation when some employee is leaving the company. Then we remove his emailaddress but we don&#039;t remove the imap account immediately.

Would it be possible to change the sql layout yet?

**Comment by Gloirin** [comment]
**Comment posted by raghnall on 20 Aug 2010 08:33**

Actually the email column in the SQL database is solely for Tine 2.0, It doesn&#039;t actually ever get used by the IMAP or SMTP server. You could make the email section null if you would like, actually it would probably be best to set DEFAULT NULL for the SQL table.

The way the SMTP and IMAP work with each other is as such.
For IMAP/SMTP SALS Auth, the username field is queried and returns the password  with the password scheme concated to it. 
For Mail delivery: 
The virtual domain is check to exist from dovecot_aliases.source
The virtual mailbox is checked to exist from dovecot_users.username
The Alias maps are checked from dovecot_aliases.source

The mail is stored based off of the user name, not the e-mail address. This prevents the accidental creation of multiple mailboxes for a single user if you should, for some reason, change the email address. 

The mail is only delivered by postfix if there exists an email -&gt; email alias. When an email account is created in tine, the driver automatically creates the username -&gt; username alias. It also removes it from the list from showing up on the SMTP settings panel. If the email address is something different than the username, the driver also creates an email -&gt; username alias. As with these defaults they are removed from showing up in the SMTP settings panel.

If you are looking for a way to disable an account without deleting it, the forwards_only could be used, just modify the dovecot Auth query to only check for username. The only side effect this has is that it will remove all aliases that are not forwards, but it does get rid of the default username -&gt; username mapping as well as the possible email -&gt; username mapping.

**Comment by Gloirin** [comment]
**Comment posted by raghnall on 20 Aug 2010 08:37**

I should clarify this:

For Mail delivery - Posfix:
The virtual domain is check to exist from dovecot_aliases.source
The virtual mailbox is checked to exist from dovecot_users.username
The Alias maps are checked from dovecot_aliases.source

Postfix then passes the mail to the Dovecot LDA (MTA)
For Mail delivery - Dovecot:
queries dovecot_users.username and returns the UID, GID, and home directory for actual mail storage.

Also on a side note: 
In order for system mail to work correctly with Felamimail users must authenticate using the username instead of the email address. 

We/You/I could modify the SMTP driver to use the email address and only make the email -&gt; username mapping in the alias table. 
The alias table contains all alias mappings (aliases and forwards) to reduce the amount of SQL queries postfix must perform.
As things stand the Dovecot SMTP driver is just a way to make postfix more efficient. The Dovecot IMAP driver could be paired with the built in Postfix MySQL driver without any problems.

**Comment by Gloirin** [comment]
**Comment posted by raghnall on 20 Aug 2010 10:56**

After some deliberation I have decided to pull the Dovecot SMTP and just use the Postfix MySQL Driver.
Doing this makes setting up Postfix a little more difficult, but I have included the new instructions that should allow it to work correctly. Please note that the instructions to get Postfix working with the Postfix driver are currently untested.

The updated Dovecot IMAP driver contains the following files(2010-8-20):
./Setup/js/EmailPanel.js
./Tinebase/EmailUser.php
./Tinebase/EmailUser/Imap/Dovecot.php
./Tinebase/EmailUser/Smtp/Postfix.php

Side note: Postfix driver needs to set username to include domain for this to work correctly. I have supplied a modified Postfix Driver in the file as well.
Changes are on lines 114, 152, 473-475, 511-524

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 20 Aug 2010 12:00**

Good idea! 

I&#039;ll review your changes. Maybe I can simplify the Postfix MySQL backend.

**Comment by Gloirin** [comment]
**Comment posted by raghnall on 20 Aug 2010 19:34**

If you would like I could use my alias/forward code to make it easier on the Postfix side. It would mean a little more complex php code, but saves SQL queries every time mail is delivered.

**Comment by Gloirin** [comment]
**Comment posted by raghnall on 21 Aug 2010 00:16**

OK a new Dovecot IMAP driver again. I have fixed a few issues with using it alongside the Postfix driver (or any other SMTP driver for that matter).

I have included a modified version of the Postfix driver that save the aliases and forwards in a single table. 
It also makes a default email -&gt; username alias (for delivery purposes) as well as a username -&gt; username alias if the email != username (again for delivery purposes), this makes the SMTP server configuration significantly easier. 

The modifies Postfix driver could just as easily be renamed to the Dovecot smtp driver if you do not wish to replace the current postfix implementation.
One final note, do not use the modified postfix driver from 2010-8-20, it doesn&#039;t work right and introduced bugs that it didn&#039;t need.

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 21 Aug 2010 11:33**

I have no problem using the updated postfix mysql backend from you.

I&#039;ll review your changes and will have a look at the dbmail imap mysql driver. If everything goes well, we will have a generic postfx mysql backend and specific dbmail/dovecot imap backends which all work together.

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 16 Sep 2010 09:49**

I put the dovecot imap backend into svn. I changed the flow and the sql layout a little bit.

Next I will merge the Posfix SMTP backend classes.

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 5 Dec 2010 19:18**

Everything is in SVN now and unit tests are written.

Thanks for your initial code.
